wom/src/main/scala/wom/expression/WomExpression.scala

```diff
@@ -14,13 +20,33 @@ trait WomExpression {
   def evaluateValue(inputValues: Map[String, WdlValue], ioFunctionSet: IoFunctionSet): ErrorOr[WdlValue]
   def evaluateType(inputTypes: Map[String, WdlType]): ErrorOr[WdlType]
   def evaluateFiles(inputTypes: Map[String, WdlValue], ioFunctionSet: IoFunctionSet, coerceTo: WdlType): ErrorOr[Set[WdlFile]]
+  def linkWithInputs(graphNodeSetter: GraphNode.GraphNodeSetter, inputMapping: Map[String, OutputPort]): ErrorOr[InstantiatedExpression] = {
+
+    def linkInput(input: String): ErrorOr[(String, InputPort)] = if (inputMapping.contains(input)) {
+      val upstreamPort = inputMapping(input)
+      Valid((input, ConnectedInputPort(input, upstreamPort.womType, upstreamPort, graphNodeSetter.get)))
+    } else {
+      s"Expression cannot be connected without the input $input (provided: ${inputMapping.toString})".invalidNel
+    }
+
+    import lenthall.validation.ErrorOr.ShortCircuitingFlatMap
+    for {
+      linkedInputList <- inputs.toList traverse linkInput
+      linkedInputs = linkedInputList.toMap
+      inputTypes = linkedInputs map { case (k, v) => k -> v.womType }
+      evaluatedType <- evaluateType(inputTypes)
+    } yield new InstantiatedExpression(this, evaluatedType, linkedInputs)
+  }
 }
 
 final case class PlaceholderWomExpression(inputs: Set[String], fixedWomType: WdlType) extends WomExpression {
   override def sourceString: String = "placeholder"
-  override def evaluateValue(inputValues: Map[String, WdlValue], ioFunctionSet: IoFunctionSet): ErrorOr[WdlValue] = Valid(WdlString("42"))
```

Comment:
Did this not break tests?

and if this didn't... are we actually using `PlaceholderWomExpression` anywhere any more?